antsibull/docsite_docs.py

```diff
@@ -0,0 +1,226 @@
+    '''Load a docsite extra-docs.yml file.
+
+    :arg index_path: Path to extra-docs.yml (does not need to exist).
+    :arg report_not_existing: Whether to report when extra-docs.yml does not exist.
+    :returns: A tuple consisting of a list of sections and a list of error messages.
+    '''
+    sections: t.List[Section] = []
+    errors: t.List[str] = []
+
+    if os.path.isfile(index_path):
+        try:
+            index = load_yaml_file(index_path)
+            if index.get('sections'):
+                for section_index, yaml_section in enumerate(index['sections']):
+                    if not isinstance(yaml_section, dict):
+                        errors.append(f'Section #{section_index} must be a mapping')
+                        continue
+                    section, section_errors = load_section(yaml_section, section_index)
+                    if section is not None:
+                        sections.append(section)
+                    errors.extend(section_errors)
+        except Exception as e:
+            errors.append(str(e))
+    elif report_not_existing:
+        errors.append('extra-docs.yml does not exist')
+
+    return sections, errors
+
+
+async def load_collection_docsite(collection_name: str,
+                                  collection_metadata: AnsibleCollectionMetadata,
+                                  path_prefix: str = 'docsite/'
+                                  ) -> CollectionDocsiteInfoT:
+    '''Given a collection name and collection metadata, load docsite data.
+
+    :arg collection_name: Dotted collection name.
+    :arg collection_metadata: Collection metadata object.
+    :arg path_prefix: Prefix to add to relative paths, and toctree entries.
+    :returns: A tuple consisting of a list of sections and a list of RST documents as tuples
+              (relative path in docsite, content).
+    '''
+    flog = mlog.fields(func='load_collection_docsite')
+    flog.debug('Enter')
+
+    index_path = os.path.join(collection_metadata.path, 'docs', 'docsite', 'extra-docs.yml')
+    sections, _ = load_docsite_index(index_path)
+    for section in sections:
+        for i, toctree in enumerate(section.toctree):
+            section.toctree[i] = path_prefix + toctree
+    documents = []
+    for doc in find_docsite_docs(collection_metadata.path):
+        try:
+            # Load content
+            async with aiofiles.open(doc[0], 'r', encoding='utf-8') as f:
+                content = await f.read()
+
+            # Lint content
+            _, errors = lint_required_conditions(content, collection_name)
+
+            # When no errors were found, add to output
+            if not errors:
+                documents.append((path_prefix + doc[1], content))
+        except Exception:
+            pass
+
+    flog.debug('Leave')
+    return sections, documents
+
+
+async def load_collections_docsites(collection_metadata: t.Mapping[str, AnsibleCollectionMetadata]
```

Comment:
I notice that we're only using `collection_name` and `collection_metadata.path` from this data structure in this function.  In my vision for the code base, I'd like there to be a split between "application" level code and "library" level code.

Functions in the library code should try to do specific things with the minimum information necessary to do its job.

Functions in the application code can do more aggregated tasks with messy data being passed around.

I want to put all application code in cli or a subdirectory of cli (`cli/doc_commands`, for instance.).

I'm not sure precisely where the line should be drawn in this piece of code.  Maybe `load_collections_docsites()` should be application level code and live under cli/doc_commands somewhere and `load_collection_docsite()` would be library level code that takes the collection_name and path as input parameters?

Reply:
I've reduced the data passed to just the collection path in 4f222971f8e07fb7245e8a5b643633654ddcd5bc.